src/coreclr/jit/optimizer.cpp

```diff
@@ -9041,12 +9041,742 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+        //        B1: brtrue(!t1, B3)
+        //        B2: return(t2)
+        //        B3: return(false)
+        //    we will try to fold it to :
+        //        B1: return(t1|t2)
+        if (b1->bbJumpDest != b2->bbNext)
+        {
+            // b1->bbJumpDest->bbNum == n1 + 2
+            return nullptr;
+        }
+    }
+
+    // Find the block conditions of b1 and b2
+
+    if (b2->countOfInEdges() > 1 || (optReturnBlock && b3->countOfInEdges() > 1))
+    {
+        return nullptr;
+    }
+
+    // Find the condition for the first block
+
+    Statement* s1 = b1->lastStmt();
+
+    GenTree* t1 = s1->GetRootNode();
+    noway_assert(t1->gtOper == GT_JTRUE);
+
+    // The second and the third block must contain a single statement
+
+    Statement* s2 = b2->firstStmt();
+    if (s2->GetPrevStmt() != s2)
+    {
+        return nullptr;
+    }
+
+    GenTree* t2 = s2->GetRootNode();
+
+    if (!optReturnBlock)
+    {
+        noway_assert(t2->gtOper == GT_JTRUE);
+    }
+    else
+    {
+        if (t2->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        Statement* s3 = b3->firstStmt();
+        if (s3->GetPrevStmt() != s3)
+        {
+            return nullptr;
+        }
+
+        GenTree* t3 = s3->GetRootNode();
+        if (t3->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        if (!varTypeIsIntegral(t2->TypeGet()) || !varTypeIsIntegral(t3->TypeGet()))
+        {
+            return nullptr;
+        }
+
+        // The third block is Return with "CNS_INT int 0/1"
+        if (t3->AsOp()->gtOp1->gtOper != GT_CNS_INT)
+        {
+            return nullptr;
+        }
+
+        if (t3->AsOp()->gtOp1->gtType != TYP_INT)
+        {
+            return nullptr;
+        }
+
+        pOptBoolsDsc->t3 = t3;
+    }
+
+    pOptBoolsDsc->t1 = t1;
+    pOptBoolsDsc->t2 = t2;
+
+    return s1;
+}
+
+//
+// optOptimizeBoolsChkTypeCostCond: Checks if type conditions meet the folding condition, and
+//                                  if cost to fold is not too expensive.
+//
+// Arguments:
+//      pOptBoolsDsc    The descriptor for boolean optimization with pointers to tree compare nodes
+//                      and its first operand, where
+//                          t1CompPtr:  The first tree pointing to GT_NE or GT_EQ GenTree node
+//                          t2CompPtr:  The second tree pointing to GT_NE or GT_EQ GenTree node
+//                          c1:         The first operand of t1CompPtr
+//                          c2:         The first operand of t2CompPtr
+
+// Return:
+//      True if it meets type conditions and cost conditions.	Else false.
+//
+bool Compiler::optOptimizeBoolsChkTypeCostCond(OptBoolsDsc* pOptBoolsDsc)
+{
+    GenTree* t1CompPtr = pOptBoolsDsc->t1CompPtr;
+    GenTree* t2CompPtr = pOptBoolsDsc->t2CompPtr;
+    GenTree* c1        = pOptBoolsDsc->c1;
+    GenTree* c2        = pOptBoolsDsc->c2;
```

Comment:
You might recall we talked about first moving the data to a struct and then possibly moving the methods too.

Code like this would possibly have a more natural feeling if these methods were members of `OptBoolDsc` rather than of `Compiler` as you wouldn't necessarily copy from the struct to locals like this but just operate on the struct members.

However other operations (that require using the compiler object) would end up being a bit more verbose.

Reply:
Will work on it.

Reply:
Done.